src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,116 @@
+ * for future use on the same attribute path.
+ */
+class AttributeValueEncoder
 {
 public:
-    AttributeValueEncoder(TLV::TLVWriter * aWriter, FabricIndex aAccessingFabricIndex) :
-        TagBoundEncoder(aWriter, TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData))),
-        mAccessingFabricIndex(aAccessingFabricIndex)
+    class ListEncodeHelper
+    {
+    public:
+        ListEncodeHelper(AttributeValueEncoder & encoder) : mAttributeValueEncoder(&encoder) {}
+
+        template <typename... Ts>
+        CHIP_ERROR Encode(Ts... aArgs) const
+        {
+            return mAttributeValueEncoder->EncodeListItem(std::forward<Ts>(aArgs)...);
+        }
+
+    private:
+        AttributeValueEncoder * mAttributeValueEncoder;
+    };
+
+    class AttributeEncodeState
+    {
+    public:
+        AttributeEncodeState() = default;
+
+        bool mAllowPartialData            = false;
+        ListIndex mCurrentEncodeListIndex = kInvalidListIndex;
+    };
+
+    AttributeValueEncoder(AttributeReportIB::Builder & aAttributeReportIBBuilder, FabricIndex aAccessingFabricIndex,
+                          const ConcreteAttributePath & aPath, const AttributeEncodeState & aState = { false, kInvalidListIndex }) :
+        mAttributeReportIBBuilder(aAttributeReportIBBuilder),
+        mAccessingFabricIndex(aAccessingFabricIndex), mPath(aPath), mEncodeState(aState)
     {}
 
     template <typename... Ts>
     CHIP_ERROR Encode(Ts... aArgs)
     {
         mTriedEncode = true;
-        if (mWriter == nullptr)
+
+        AttributeDataIB::Builder attributeDataIBBuilder = mAttributeReportIBBuilder.CreateAttributeData();
+        AttributePathIB::Builder attributePathIBBuilder = attributeDataIBBuilder.CreatePath();
+        attributePathIBBuilder.Endpoint(mPath.mEndpointId).Cluster(mPath.mClusterId).Attribute(mPath.mAttributeId);
+
+        if (mCurrentEncodingListIndex != kInvalidListIndex)
+        {
+            attributePathIBBuilder.ListIndex(mCurrentEncodingListIndex);
+        }
+
+        attributePathIBBuilder.EndOfAttributePathIB();
+
+        ReturnErrorOnFailure(
+            TagBoundEncoder(attributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)))
+                .Encode(std::forward<Ts>(aArgs)...));
+
+        attributeDataIBBuilder.DataVersion(0).EndOfAttributeDataIB();
+        return attributeDataIBBuilder.GetError();
+    }
+
+    template <typename... Ts>
+    CHIP_ERROR EncodeListItem(Ts... aArgs)
+    {
+        if (mEncodeState.mCurrentEncodeListIndex == kInvalidListIndex)
         {
+            AttributeDataIB::Builder attributeDataIBBuilder = mAttributeReportIBBuilder.CreateAttributeData();
+            attributeDataIBBuilder.CreatePath()
+                .Endpoint(mPath.mEndpointId)
+                .Cluster(mPath.mClusterId)
+                .Attribute(mPath.mAttributeId)
+                .EndOfAttributePathIB();
+
+            ReturnErrorOnFailure(
```

Comment:
Comments please a) as to the fact that we're encoding an empty list here and b) why we're encoding an empty list.